jdaviz/app.py

```diff
@@ -2022,7 +2115,37 @@ def set_data_visibility(self, viewer_reference, data_label, visible=True, replac
                 viewer.on_limits_change()  # Trigger compass redraw
 
     def vue_data_item_remove(self, event):
-        self.data_collection.remove(self.data_collection[event['item_name']])
+
+        data_label = event['item_name']
+        data = self.data_collection[data_label]
+        self._reparent_subsets(data)
+
+        # Imviz has some extra logic below that can be skipped after data removal if we're not
+        # removing the reference data, so we check that here.
+        if self.config == "imviz":
+            imviz_refdata = False
+            ref_data, iref = self._jdaviz_helper.get_ref_data()
+            if data is ref_data:
+                imviz_refdata = True
+
+        self.data_collection.remove(self.data_collection[data_label])
+
+        # If there are two or more datasets left we need to link them back together after removing
+        # the reference data (which would leave 0 external_links).
+        if len(self.data_collection) > 1 and len(self.data_collection.external_links) == 0:
+            if self.config == "imviz" and imviz_refdata:
+                link_type = self._jdaviz_helper.plugins["Links Control"].link_type.selected.lower()
+                self._jdaviz_helper.link_data(link_type=link_type, error_on_fail=True)
+                # Hack to restore responsiveness to imviz layers
+                for viewer_ref in self.get_viewer_reference_names():
+                    viewer = self.get_viewer(viewer_ref)
+                    loaded_layers = [layer.layer.label for layer in viewer.layers if
+                                     "Subset" not in layer.layer.label]
+                    self.remove_data_from_viewer(viewer_ref, loaded_layers[-1])
+                    self.add_data_to_viewer(viewer_ref, loaded_layers[-1])
+            else:
+                for i in range(1, len(self.data_collection)):
+                    self._link_new_data(data_to_be_linked=i)
```

Comment:
Won't this be very slow? Don't we want to collect all the indices first before relinking them all at once?

Reply:
`_link_new_data` only links one dataset to the ref data, it can't do them all at once.

Reply:
What do you think to make it just return the links but not actually do the linking until all the links are collected?

Reply:
There is already a `link_image_data` that does this but I guess you want it to be viz-agnostic here?

Reply:
(oh, nvm, see I this this logic block is specifically for non-imviz, but I still think the more correct thing to do is to collect the links first and then set them all at once in the end)

Reply:
Yes I was going to say, that's why this is in the `not imviz` block 🙂. I'm open to doing this a faster way, this was just the function that was already available to do it. Although in my experience testing this, the re-linking in Specviz happens very quickly, it's Imviz that is very slow.